File: scripts/receive_process.py
# coding=utf-8
import rospy  # ROS库
from sensor_msgs.msg import Image   # ROS中接收传感器的图片格式，要注意的是别漏了.msg，否则导入失败
import cv_bridge    # 用于方便实现OpenCV与ROS的类型转换
import cv2  # OpenCV库


# 相关内容在之前OpenCV笔记中已经说过，此处不再赘述
def imgStretch(cv_img):
    img_hsv = cv2.cvtColor(cv_img, cv2.COLOR_BGR2HSV)
    img_h = img_hsv[:, :, 0]
    img_s = img_hsv[:, :, 1]
    img_v = img_hsv[:, :, 2]
    equal_img_v = cv2.equalizeHist(img_v)
    img_hsv[:, :, 2] = equal_img_v
    equal_img = cv2.cvtColor(img_hsv, cv2.COLOR_HSV2BGR)
    return equal_img


# 接收到数据后的回调函数
def callback(img):
    # 新建一个cv_bridge对象
    bridge = cv_bridge.CvBridge()

    # 调用bridge对象的成员函数imgmsg_to_cv2将ROS的Image转成OpenCV的NDArray
    # 第一个参数是待转换数据
    # 第二个参数是编码格式字符串，在sensor_msgs/image_encodings.h中有定义
    cv_img = bridge.imgmsg_to_cv2(img, "bgr8")

    # 调用直方图均衡化函数对彩色图像进行拉伸
    equal_img = imgStretch(cv_img)

    # 再次调用bridge的成员函数cv2_to_imgmsg将NDArray类型转成ROS的Image类型
    # 第一个参数是待转换数据
    # 第二个参数是待编码格式字符串，这里因为还是8位彩色图像，所以不变
    img_msg = bridge.cv2_to_imgmsg(equal_img, "bgr8")

    # 调用发布函数发布数据
    pub.publish(img_msg)

    # 调用OpenCV展示直接接收到的原始数据
    cv2.imshow("data from camera", cv_img)
    cv2.waitKey(3)


if __name__ == "__main__":
    # 初始化节点并指定名称
    rospy.init_node("receive_process")

    # 新建Subscriber用于接收原始数据
    # 第一个参数是接收Topic的名称
    # 第二个参数是接收数据的类型
    # 第三个参数是接收到数据后的回调函数
    rospy.Subscriber("image_raw", Image, callback)

    # 新建Publisher用于发布处理后的数据
    # 第一个参数是发布的Topic名称
    # 第二个参数是发布的数据类型
    # 第三个参数是队列长度，一般是一个较小的整数即可
    pub = rospy.Publisher("processed_img", Image, queue_size=1)

    # 循环执行
    rospy.spin()

File: src/receive_process.cpp
# include<ros/ros.h>    // ROS头文件
# include<cv_bridge/cv_bridge.h>    // cv_bridge头文件，实现类型转换
# include<sensor_msgs/Image.h>  // ROS相机传感器数据格式
# include<opencv-3.3.1-dev/opencv2/opencv.hpp>  //OpenCV核心头文件

// 定义的全局变量
ros::Subscriber sub;
ros::Publisher pub;

// 对彩色图像进行灰度直方图均衡化，增强对比度
cv::Mat imgStretch(cv::Mat cv_img){
    cv::Mat img_hsv;
    cv::cvtColor(cv_img,img_hsv,cv::COLOR_BGR2HSV);
    std::vector<cv::Mat> hsv_bands;
    cv::split(img_hsv,hsv_bands);
    cv::equalizeHist(hsv_bands[2],hsv_bands[2]);
    cv::Mat equal_img;
    cv::merge(hsv_bands,equal_img);
    cv::cvtColor(equal_img,equal_img,cv::COLOR_HSV2BGR);
    return equal_img;
}

cv::Mat cvtImgMsg2Mat(const sensor_msgs::ImageConstPtr& msg){
    // 利用cv_bridge::toCvCopy函数将消息格式的图片转成CV格式
    cv_bridge::CvImagePtr cvImgPtr;
    cvImgPtr = cv_bridge::toCvCopy(msg,sensor_msgs::image_encodings::BGR8);
    return cvImgPtr->image;
}

sensor_msgs::ImagePtr cvtMat2ImgMsg(cv::Mat img){
    // 利用cv_bridge::CvImage函数将CV图像再次转成消息格式
    sensor_msgs::ImagePtr img_msg;
    img_msg = cv_bridge::CvImage(std_msgs::Header(),"bgr8",img).toImageMsg();
    return img_msg;
}

void callback(const sensor_msgs::ImageConstPtr& msg){
    // 调用函数，将msg格式img转成mat
    cv::Mat img = cvtImgMsg2Mat(msg);

    // 对转换后的数据进行处理
    cv::Mat equal_img = imgStretch(img);
    
    // 将处理后的数据重新转成msg格式
    sensor_msgs::ImagePtr img_msg = cvtMat2ImgMsg(equal_img);

    // 发送处理后的数据
    pub.publish(img_msg);
}

int main(int argc, char *argv[])
{
    // 初始化节点并指定名称
    ros::init(argc, argv, "receive_process");
    // 新建节点管理器
    ros::NodeHandle nh;

    // 初始化pub和sub
    sub = nh.subscribe("image_raw",1,callback);
    pub = nh.advertise<sensor_msgs::Image>("processed_img",1);
    
    // 循环执行
    ros::spin();
    return 0;
}